File: java/src/main/java/io/github/c9jimmy/mayancalc/Calculator.java
package io.github.c9jimmy.mayancalc;

public final class Calculator {
    private Calculator() {}

    /** Gregorian date → Julian Day Number (Meeus algorithm). */
    static int dateToJdn(int year, int month, int day) {
        int y = (month <= 2) ? year - 1 : year;
        int m = (month <= 2) ? month + 12 : month;
        int a = y / 100;
        int b = 2 - a + a / 4;
        return (int)(Constants.MEEUS_YEAR_FACTOR * (y + Constants.MEEUS_EPOCH_A)) + (int)(Constants.MEEUS_MONTH_FACTOR * (m + 1)) + day + b - Constants.MEEUS_EPOCH_B;
    }

    /**
     * JDN → Tzolk'in date.
     *
     * Math.floorMod handles pre-epoch dates (total < 0) correctly.
     * Tzolk'in origin constants align creation date (0.0.0.0.0) with 4 Ajaw.
     */
    static TzolkinDate jdnToTzolkin(int jdn) {
        int kin = Math.floorMod(jdn - Constants.GMT_CORRELATION, Constants.TZOLKIN_CYCLE);
        int coefficient = ((kin + Constants.TZOLKIN_COEFF_ORIGIN - 1) % Constants.TZOLKIN_COEFF_COUNT) + 1;
        int signIdx = (kin + Constants.TZOLKIN_NAME_ORIGIN_INDEX) % Constants.TZOLKIN_SIGN_COUNT;
        return new TzolkinDate(coefficient, Constants.TZOLKIN_DAY_SIGNS.get(signIdx), signIdx + 1);
    }

    /**
     * JDN → Haab date.
     * +348 aligns creation date with 8 Kumk'u (position 17×20+8 = 348).
     */
    static HaabDate jdnToHaab(int jdn) {
        int haabKin = Math.floorMod(jdn - Constants.GMT_CORRELATION + Constants.HAAB_CORRELATION_OFFSET, Constants.HAAB_CYCLE);
        return new HaabDate(haabKin % Constants.HAAB_DAYS_PER_MONTH, Constants.HAAB_MONTHS.get(haabKin / Constants.HAAB_DAYS_PER_MONTH));
    }

    /** JDN → Long Count (baktun.katun.tun.uinal.kin). */
    static LongCount jdnToLongCount(int jdn) {
        int total = jdn - Constants.GMT_CORRELATION;
        return new LongCount(
            total / Constants.LC_BAKTUN,
            (total % Constants.LC_BAKTUN) / Constants.LC_KATUN,
            (total % Constants.LC_KATUN)  / Constants.LC_TUN,
            (total % Constants.LC_TUN)    / Constants.LC_UINAL,
            total % Constants.LC_UINAL
        );
    }

    /** JDN → Lord of Night G1–G9 (9-day cycle). */
    static String jdnToLordOfNight(int jdn) {
        int total = jdn - Constants.GMT_CORRELATION;
        int lord = Math.floorMod(total + Constants.LORD_OF_NIGHT_ORIGIN - 1, Constants.LORD_OF_NIGHT_CYCLE) + 1;
        return "G" + lord;
    }

    private static boolean isLeapYear(int year) {
        return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
    }

    private static int daysInMonth(int year, int month) {
        int[] monthLengths = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
        if (month == 2 && isLeapYear(year)) {
            return 29;
        }
        return monthLengths[month - 1];
    }

    private static void validateDate(int year, int month, int day) {
        if (month < 1 || month > 12) {
            throw new IllegalArgumentException("month must be in 1..12");
        }
        int maxDay = daysInMonth(year, month);
        if (day < 1 || day > maxDay) {
            throw new IllegalArgumentException("day must be in 1.." + maxDay + " for month " + month);
        }
    }

    /** Gregorian date → complete Classic Maya calendar output. */
    public static MayanDate calculate(int year, int month, int day) {
        validateDate(year, month, day);
        int jdn = dateToJdn(year, month, day);
        return new MayanDate(
            jdnToTzolkin(jdn),
            jdnToHaab(jdn),
            jdnToLongCount(jdn),
            jdnToLordOfNight(jdn)
        );
    }
}

File: kotlin/src/main/kotlin/io/github/c9jimmy/mayancalc/Calculator.kt
package io.github.c9jimmy.mayancalc

private fun floorMod(a: Long, b: Long): Long = ((a % b) + b) % b

object Calculator {

    internal fun dateToJdn(year: Int, month: Int, day: Int): Long {
        val y = if (month <= 2) year - 1 else year
        val m = if (month <= 2) month + 12 else month
        val a = y / 100
        val b = 2 - a + a / 4
        return (MEEUS_YEAR_FACTOR * (y + MEEUS_EPOCH_A)).toLong() +
               (MEEUS_MONTH_FACTOR * (m + 1)).toLong() +
               day + b - MEEUS_EPOCH_B
    }

    internal fun jdnToTzolkin(jdn: Long): TzolkinDate {
        val kin = floorMod(jdn - GMT_CORRELATION, TZOLKIN_CYCLE).toInt()
        val coefficient = (kin + TZOLKIN_COEFF_ORIGIN - 1) % TZOLKIN_COEFF_COUNT + 1
        val nameIdx = (kin + TZOLKIN_NAME_ORIGIN_INDEX) % TZOLKIN_SIGN_COUNT
        return TzolkinDate(coefficient, TZOLKIN_DAY_SIGNS[nameIdx], nameIdx + 1)
    }

    internal fun jdnToHaab(jdn: Long): HaabDate {
        val haabKin = floorMod(jdn - GMT_CORRELATION + HAAB_CORRELATION_OFFSET, HAAB_CYCLE).toInt()
        val monthIdx = haabKin / HAAB_DAYS_PER_MONTH
        return HaabDate(haabKin % HAAB_DAYS_PER_MONTH, HAAB_MONTHS[monthIdx])
    }

    internal fun jdnToLongCount(jdn: Long): LongCount {
        val total = jdn - GMT_CORRELATION
        return LongCount(
            (total / LC_BAKTUN).toInt(),
            ((total % LC_BAKTUN) / LC_KATUN).toInt(),
            ((total % LC_KATUN) / LC_TUN).toInt(),
            ((total % LC_TUN) / LC_UINAL).toInt(),
            (total % LC_UINAL).toInt()
        )
    }

    internal fun jdnToLordOfNight(jdn: Long): String {
        val total = jdn - GMT_CORRELATION
        val g = floorMod(total + LORD_OF_NIGHT_ORIGIN - 1, LORD_OF_NIGHT_CYCLE).toInt() + 1
        return "G$g"
    }

    private fun isLeapYear(year: Int): Boolean =
        year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)

    private fun daysInMonth(year: Int, month: Int): Int {
        val monthLengths = intArrayOf(31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31)
        return if (month == 2 && isLeapYear(year)) 29 else monthLengths[month - 1]
    }

    private fun validateDate(year: Int, month: Int, day: Int) {
        require(month in 1..12) { "month must be in 1..12" }
        val maxDay = daysInMonth(year, month)
        require(day in 1..maxDay) { "day must be in 1..$maxDay for month $month" }
    }

    fun calculate(year: Int, month: Int, day: Int): MayanDate {
        validateDate(year, month, day)
        val jdn = dateToJdn(year, month, day)
        return MayanDate(
            jdnToTzolkin(jdn),
            jdnToHaab(jdn),
            jdnToLongCount(jdn),
            jdnToLordOfNight(jdn)
        )
    }
}
